README.md
# isomorphic-timers-promises

[![Build Status][ci-img]][ci]
[![BrowserStack Status][browserstack-img]][browserstack]

[`timers/promises`](https://nodejs.org/api/timers.html#timers_timers_promises_api)
for client and server.

> The `timers/promises` API provides an alternative set of timer functions that
> return `Promise` objects.

## Install

```sh
npm install isomorphic-timers-promises --save
```

## Usage

```js
import {
	setTimeout,
	setImmediate,
	setInterval
} from 'isomorphic-timers-promises';

(async () => {
	const result = await setTimeout(100, 'becky');
	console.log(result); // 'becky'
})();

(async () => {
	const result = await setImmediate('maya');
	console.log(result); // 'maya'
})();

(async () => {
	let result = 0;
	for await (const startTime of setInterval(100, Date.now())) {
		const now = Date.now();
		result = result + 1;
		if (now - startTime >= 1000) {
			break;
		}
	}
	console.log(result); // 10
})();
```

### Usage with Webpack

<details>
	
<summary>Show me</summary>

```js
// webpack.config.js
module.exports = {
	// ...
	resolve: {
		alias: {
			'timers/promises': 'isomorphic-timers-promises'
		}
	}
};
```

</details>

### Usage with Rollup

<details>
	
<summary>Show me</summary>

```js
// rollup.config.js
const { default: resolve } = require('@rollup/plugin-node-resolve');
const alias = require('@rollup/plugin-alias');

module.exports = {
	// ...
	plugins: [
		resolve(),
		alias({
			entries: {
				'timers/promises': 'isomorphic-timers-promises'
			}
		})
	]
};
```

</details>

### Usage in Node

<details>
	
<summary>Show me</summary>

Depending on your configuration you can alias `timers/promises` with packages
like [`link-module-alias`](https://github.com/Rush/link-module-alias) or
[`babel-plugin-module-resolver`](https://github.com/tleunen/babel-plugin-module-resolver).

</details>

## API

### setTimeout([delay[, value[, options]]])

Returns: `Promise`

| Property         | Type          | Default | Description                                                                                                                                  |
| ---------------- | ------------- | ------- | -------------------------------------------------------------------------------------------------------------------------------------------- |
| `delay`          | `number`      | `1`     | The number of milliseconds to wait before fulfilling the promise.                                                                            |
| `value`          | `*`           |         | A value with which the promise is fulfilled.                                                                                                 |
| `options.ref`    | `boolean`     | `true`  | Set to `false` to indicate that the scheduled timeout should not require the event loop to remain active. Valid only for server environment. |
| `options.signal` | `AbortSignal` |         | An optional `AbortSignal` that can be used to cancel the scheduled timeout.                                                                  |

### setImmediate([value[, options]])

Returns: `Promise`

| Property         | Type          | Default | Description                                                                                                                                    |
| ---------------- | ------------- | ------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
| `value`          | `*`           |         | A value with which the promise is fulfilled.                                                                                                   |
| `options.ref`    | `boolean`     | `true`  | Set to `false` to indicate that the scheduled immediate should not require the event loop to remain active. Valid only for server environment. |
| `options.signal` | `AbortSignal` |         | An optional `AbortSignal` that can be used to cancel the scheduled immediate.                                                                  |

### setInterval([delay[, value[, options]]])

Returns: async iterator that generates values in an interval of `delay`.

| Property         | Type          | Default | Description                                                                                                                                                     |
| ---------------- | ------------- | ------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `delay`          | `number`      | `1`     | The number of milliseconds to wait between iterations.                                                                                                          |
| `value`          | `*`           |         | A value with which the iterator returns.                                                                                                                        |
| `options.ref`    | `boolean`     | `true`  | Set to `false` to indicate that the scheduled timeout between iterations should not require the event loop to remain active. Valid only for server environment. |
| `options.signal` | `AbortSignal` |         | An optional `AbortSignal` that can be used to cancel the scheduled timeout between operations.                                                                  |

## Node and browser support

Supports Node 10+.

Tested in Chrome 72, Firefox 65, Internet Explorer 11 and should work in all
modern browsers.

Check
[support based on Browserslist configuration](https://browserslist.dev/?q=bGFzdCAzIG1ham9yIHZlcnNpb25zLCBzaW5jZSAyMDE5LCBub3QgaWUgPD0gMTAsIG5vZGUgMTA%3D)).

Assumes `Promise`, `AbortController` and `setImmediate` are polyfilled or
available in global context.

## Test

For automated tests, run `npm run test:automated` (append `:watch` for watcher
support).

Test suite is taken and modified from official Node.js repository
([`setTimeout`](https://github.com/nodejs/node/blob/master/test/parallel/test-timers-timeout-promisified.js),
[`setImmediate`](https://github.com/nodejs/node/blob/master/test/parallel/test-timers-immediate-promisified.js),
[`setInterval`](https://github.com/nodejs/node/blob/master/test/parallel/test-timers-interval-promisified.js)).

## License

MIT © [Ivan Nikolić](http://ivannikolic.com)

<!-- prettier-ignore-start -->

[ci]: https://travis-ci.com/niksy/isomorphic-timers-promises
[ci-img]: https://travis-ci.com/niksy/isomorphic-timers-promises.svg?branch=master
[browserstack]: https://www.browserstack.com/
[browserstack-img]: https://www.browserstack.com/automate/badge.svg?badge_key=Tm5vWkh0c2l3U0lKVEtOSkxVcGFvSUs3VVZ0dHZOcHROMWtoUmlIZ1lVaz0tLUgwQ1JoSEU1YStCTi9SSTZ4RWRBRkE9PQ==--81b6442621a16459ec7e2e40dab413dff6c62aaa

<!-- prettier-ignore-end -->


CHANGELOG.md
# Changelog

## [Unreleased][]

### Added

-   Configuration for Node support

## [1.0.0][] - 2021-06-29

### Added

-   Initial implementation

[unreleased]:
	https://github.com/niksy/isomorphic-timers-promises/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/niksy/isomorphic-timers-promises/tree/v1.0.0


test/index.js
/* globals globalThis:false */

import assert from 'assert';
import setImmediateShim from 'set-immediate-shim';
import 'abortcontroller-polyfill/dist/abortcontroller-polyfill-only';
import {
	setTimeout as setTimeoutPromise,
	setImmediate as setImmediatePromise,
	setInterval as setIntervalPromise
} from '../index';

if (!globalThis.setImmediate) {
	globalThis.setImmediate = setImmediateShim;
	globalThis.clearImmediate = function () {};
}

const symbolAsyncIterator =
	Symbol?.asyncIterator ?? Symbol?.iterator ?? '@@asyncIterator';

before(function () {
	window.fixture.load('/test/fixtures/index.html');
});

after(function () {
	window.fixture.cleanup();
});

describe('setTimeout', function () {
	it('should handle case with value', async function () {
		const start = new Date();
		const result = await setTimeoutPromise(1000, 'becky');
		const end = new Date();
		assert.ok(end - start >= 1000);
		assert.equal(result, 'becky');
	});

	it('should handle case with no value', async function () {
		const result = await setTimeoutPromise(0);
		assert.equal(typeof result, 'undefined');
	});

	it('should handle aborted operation after promise', async function () {
		const controller = new AbortController();
		const start = new Date();
		const promise = setTimeoutPromise(10_000, 'becky', {
			signal: controller.signal
		});
		controller.abort();
		try {
			await promise;
		} catch (error) {
			const end = new Date();
			assert.equal(error.name, 'AbortError');
			assert.ok(end - start < 1000);
		}
	});

	it('should handle aborted operation before promise', async function () {
		const controller = new AbortController();
		controller.abort();
		const start = new Date();
		const promise = setTimeoutPromise(10_000, 'becky', {
			signal: controller.signal
		});
		try {
			await promise;
		} catch (error) {
			const end = new Date();
			assert.equal(error.name, 'AbortError');
			assert.ok(end - start < 1000);
		}
	});

	it('should not throw after resolving promise and calling abort', async function () {
		const controller = new AbortController();
		const promise = setTimeoutPromise(10, 'becky', {
			signal: controller.signal
		});
		const result = await promise;
		controller.abort();
		assert.ok(result, 'becky');
	});

	it('should validate options', async function () {
		await Promise.all(
			[1, '', false, Infinity].map(async (options) => {
				try {
					await setTimeoutPromise(1000, 'becky', options);
				} catch (error) {
					assert.equal(error.code, 'ERR_INVALID_ARG_TYPE');
				}
			})
		);
	});

	it('should validate options.signal', async function () {
		await Promise.all(
			[1, '', false, Infinity, null, {}].map(async (signal) => {
				try {
					await setTimeoutPromise(1000, 'becky', { signal });
				} catch (error) {
					assert.equal(error.code, 'ERR_INVALID_ARG_TYPE');
				}
			})
		);
	});

	it('should validate options.ref', async function () {
		await Promise.all(
			[1, '', Infinity, null, {}].map(async (reference) => {
				try {
					await setTimeoutPromise(1000, 'becky', { ref: reference });
				} catch (error) {
					assert.equal(error.code, 'ERR_INVALID_ARG_TYPE');
				}
			})
		);
	});
});

describe('setImmediate', function () {
	it('should handle case with value', async function () {
		const result = await setImmediatePromise('becky');
		assert.equal(result, 'becky');
	});

	it('should handle case with no value', async function () {
		const result = await setImmediatePromise();
		assert.equal(typeof result, 'undefined');
	});

	it('should handle aborted operation after promise', async function () {
		const controller = new AbortController();
		const start = new Date();
		const promise = setImmediatePromise('becky', {
			signal: controller.signal
		});
		controller.abort();
		try {
			await promise;
		} catch (error) {
			const end = new Date();
			assert.equal(error.name, 'AbortError');
			assert.ok(end - start < 1000);
		}
	});

	it('should handle aborted operation before promise', async function () {
		const controller = new AbortController();
		controller.abort();
		const start = new Date();
		const promise = setImmediatePromise('becky', {
			signal: controller.signal
		});
		try {
			await promise;
		} catch (error) {
			const end = new Date();
			assert.equal(error.name, 'AbortError');
			assert.ok(end - start < 1000);
		}
	});

	it('should not throw after resolving promise and calling abort', async function () {
		const controller = new AbortController();
		const promise = setImmediatePromise('becky', {
			signal: controller.signal
		});
		const result = await promise;
		controller.abort();
		assert.ok(result, 'becky');
	});

	it('should validate options', async function () {
		await Promise.all(
			[1, '', false, Infinity].map(async (options) => {
				try {
					await setImmediatePromise('becky', options);
				} catch (error) {
					assert.equal(error.code, 'ERR_INVALID_ARG_TYPE');
				}
			})
		);
	});

	it('should validate options.signal', async function () {
		await Promise.all(
			[1, '', false, Infinity, null, {}].map(async (signal) => {
				try {
					await setImmediatePromise('becky', { signal });
				} catch (error) {
					assert.equal(error.code, 'ERR_INVALID_ARG_TYPE');
				}
			})
		);
	});

	it('should validate options.ref', async function () {
		await Promise.all(
			[1, '', Infinity, null, {}].map(async (reference) => {
				try {
					await setImmediatePromise('becky', { ref: reference });
				} catch (error) {
					assert.equal(error.code, 'ERR_INVALID_ARG_TYPE');
				}
			})
		);
	});
});

describe('setInterval', function () {
	async function runInterval(function_, intervalTime, signal) {
		const input = 'becky';
		const interval = setIntervalPromise(intervalTime, input, {
			signal
		});
		let iteration = 0;
		for await (const value of interval) {
			assert.equal(value, input);
			iteration++;
			await function_(iteration);
		}
	}

	it('should handle case with value', async function () {
		const iterable = setIntervalPromise(1, 'becky');
		const iterator = iterable[symbolAsyncIterator]();
		const promise = iterator.next();

		const result = await promise;
		assert.ok(!result.done, 'iterator was wrongly marked as done');
		assert.equal(result.value, 'becky');

		return iterator.return();
	});

	it('should handle case with no value', async function () {
		const iterable = setIntervalPromise(0);
		const iterator = iterable[symbolAsyncIterator]();
		const promise = iterator.next();

		const result = await promise;
		assert.ok(!result.done, 'iterator was wronlgy marked as done');
		assert.equal(typeof result.value, 'undefined');

		return iterator.return();
	});

	it('should handle multiple iterations', async function () {
		const iterable = setIntervalPromise(1, 'becky');
		const iterator = iterable[symbolAsyncIterator]();
		const promise = iterator.next();

		let result = await promise;
		assert.ok(!result.done, 'iterator was wrongly marked as done');
		assert.equal(result.value, 'becky');
		result = await iterator.next();
		assert.ok(!result.done, 'iterator was wrongly marked as done');
		assert.equal(result.value, 'becky');

		return iterator.return();
	});

	it('should handle aborted operation after promise', async function () {
		const controller = new AbortController();

		const iterable = setIntervalPromise(100, 'becky', {
			signal: controller.signal
		});
		const iterator = iterable[symbolAsyncIterator]();
		const promise = iterator.next();

		controller.abort();

		try {
			await promise;
		} catch (error) {
			assert.equal(error.name, 'AbortError');
		}
	});

	it('should handle aborted operation before promise', async function () {
		const controller = new AbortController();
		controller.abort();

		const iterable = setIntervalPromise(1, 'becky', {
			signal: controller.signal
		});
		const iterator = iterable[symbolAsyncIterator]();
		const promise = iterator.next();

		try {
			await promise;
		} catch (error) {
			assert.equal(error.name, 'AbortError');
		}
	});

	it('should not throw after resolving promise and calling abort', async function () {
		const controller = new AbortController();

		const iterable = setIntervalPromise(1, 'becky', {
			signal: controller.signal
		});
		const iterator = iterable[symbolAsyncIterator]();
		const promise = iterator.next();

		const result = await promise;
		controller.abort();
		assert.ok(result, 'becky');
	});

	it('checks that we call the correct amount of times', async function () {
		const controller = new AbortController();
		let loopCount = 0;
		const delay = 20;
		const timeoutLoop = runInterval(
			async () => {
				loopCount++;
				if (loopCount === 5) {
					controller.abort();
				}
				if (loopCount > 5) {
					throw new Error('ran too many times');
				}
			},
			delay,
			controller.signal
		);

		try {
			await timeoutLoop;
		} catch (error) {
			assert.equal(error.name, 'AbortError');
			assert.equal(loopCount, 5);
		}
	});

	it('checks that if we abort when we have some unresolved callbacks, we actually call them', async function () {
		const controller = new AbortController();
		const delay = 10;
		let totalIterations = 0;
		const timeoutLoop = runInterval(
			async (iterationNumber) => {
				await setTimeoutPromise(delay * 4);
				if (iterationNumber <= 2) {
					assert.equal(controller.signal.aborted, false);
				}
				if (iterationNumber === 2) {
					controller.abort();
				}
				if (iterationNumber > 2) {
					assert.equal(controller.signal.aborted, true);
				}
				if (iterationNumber > totalIterations) {
					totalIterations = iterationNumber;
				}
			},
			delay,
			controller.signal
		);

		try {
			await timeoutLoop;
		} catch (error) {
			assert.ok(
				totalIterations >= 3,
				`iterations was ${totalIterations} < 3`
			);
		}
	});

	it('checks that the timing is correct', async function () {
		let pre = false;
		let post = false;
		const timeUnit = 50;

		await Promise.all([
			(async () => {
				await setTimeoutPromise(1);
				pre = true;
			})(),
			new Promise((resolve) => {
				const iterable = setIntervalPromise(timeUnit * 2);
				const iterator = iterable[symbolAsyncIterator]();

				iterator
					.next()
					.then(() => {
						assert.ok(pre, 'interval ran too early');
						assert.ok(!post, 'interval ran too late');
						return iterator.next();
					})
					.then(() => {
						assert.ok(post, 'second interval ran too early');
						return iterator.return();
					})
					.then(resolve);
			}),
			(async () => {
				await setTimeoutPromise(timeUnit * 3);
				post = true;
			})()
		]);
	});

	it('should validate options', async function () {
		await Promise.all(
			[1, '', false, Infinity].map(async (options) => {
				const iterable = setIntervalPromise(1000, 'becky', options);
				const iterator = iterable[symbolAsyncIterator]();
				const promise = iterator.next();
				try {
					await promise;
				} catch (error) {
					assert.equal(error.code, 'ERR_INVALID_ARG_TYPE');
				}
			})
		);
	});

	it('should validate options.signal', async function () {
		await Promise.all(
			[1, '', false, Infinity, null, {}].map(async (signal) => {
				const iterable = setIntervalPromise(1000, 'becky', { signal });
				const iterator = iterable[symbolAsyncIterator]();
				const promise = iterator.next();
				try {
					await promise;
				} catch (error) {
					assert.equal(error.code, 'ERR_INVALID_ARG_TYPE');
				}
			})
		);
	});

	it('should validate options.ref', async function () {
		await Promise.all(
			[1, '', Infinity, null, {}].map(async (reference) => {
				const iterable = setIntervalPromise(1000, 'becky', {
					ref: reference
				});
				const iterator = iterable[symbolAsyncIterator]();
				const promise = iterator.next();
				try {
					await promise;
				} catch (error) {
					assert.equal(error.code, 'ERR_INVALID_ARG_TYPE');
				}
			})
		);
	});

	it('should handle example from documentation', async function () {
		let count = 0;
		for await (const startTime of setIntervalPromise(100, Date.now())) {
			const now = Date.now();
			count = count + 1;
			if (now - startTime >= 1000) {
				break;
			}
		}
		assert.equal(count, 10);
	});
});


index.js
/* eslint-disable unicorn/custom-error-definition, promise/always-return */
/* global setImmediate:false, clearImmediate:false */

const symbolAsyncIterator =
	Symbol?.asyncIterator ?? Symbol?.iterator ?? '@@asyncIterator';

class ERR_INVALID_ARG_TYPE extends Error {
	constructor(name, expected, actual) {
		super(`${name} must be ${expected}, ${typeof actual} given`);
		this.code = 'ERR_INVALID_ARG_TYPE';
	}
}

class AbortError extends Error {
	constructor() {
		super('The operation was aborted');
		this.code = 'ABORT_ERR';
	}
}

function validateObject(object, name) {
	if (object === null || typeof object !== 'object') {
		throw new ERR_INVALID_ARG_TYPE(name, 'Object', object);
	}
}

function validateBoolean(value, name) {
	if (typeof value !== 'boolean') {
		throw new ERR_INVALID_ARG_TYPE(name, 'boolean', value);
	}
}

function validateAbortSignal(signal, name) {
	if (
		typeof signal !== 'undefined' &&
		(signal === null ||
			typeof signal !== 'object' ||
			!('aborted' in signal))
	) {
		throw new ERR_INVALID_ARG_TYPE(name, 'AbortSignal', signal);
	}
}

function asyncIterator({ next: nextFunction, return: returnFunction }) {
	const result = {};
	if (typeof nextFunction === 'function') {
		result.next = nextFunction;
	}
	if (typeof returnFunction === 'function') {
		result.return = returnFunction;
	}
	result[symbolAsyncIterator] = function () {
		return this;
	};

	return result;
}

function setTimeoutPromise(after = 1, value, options = {}) {
	const arguments_ = [].concat(value ?? []);
	try {
		validateObject(options, 'options');
	} catch (error) {
		return Promise.reject(error);
	}
	const { signal, ref: reference = true } = options;
	try {
		validateAbortSignal(signal, 'options.signal');
	} catch (error) {
		return Promise.reject(error);
	}
	try {
		validateBoolean(reference, 'options.ref');
	} catch (error) {
		return Promise.reject(error);
	}
	if (signal?.aborted) {
		return Promise.reject(new AbortError());
	}
	let onCancel;
	const returnValue = new Promise((resolve, reject) => {
		const timeout = setTimeout(() => resolve(value), after, ...arguments_);
		if (!reference) {
			timeout?.unref?.();
		}
		if (signal) {
			onCancel = () => {
				clearTimeout(timeout);
				reject(new AbortError());
			};
			signal.addEventListener('abort', onCancel);
		}
	});
	if (typeof onCancel !== 'undefined') {
		returnValue.finally(() =>
			signal.removeEventListener('abort', onCancel)
		);
	}
	return returnValue;
}

function setImmediatePromise(value, options = {}) {
	try {
		validateObject(options, 'options');
	} catch (error) {
		return Promise.reject(error);
	}
	const { signal, ref: reference = true } = options;
	try {
		validateAbortSignal(signal, 'options.signal');
	} catch (error) {
		return Promise.reject(error);
	}
	try {
		validateBoolean(reference, 'options.ref');
	} catch (error) {
		return Promise.reject(error);
	}
	if (signal?.aborted) {
		return Promise.reject(new AbortError());
	}
	let onCancel;
	const returnValue = new Promise((resolve, reject) => {
		const immediate = setImmediate(() => resolve(value));
		if (!reference) {
			immediate?.unref?.();
		}
		if (signal) {
			onCancel = () => {
				clearImmediate(immediate);
				reject(new AbortError());
			};
			signal.addEventListener('abort', onCancel);
		}
	});
	if (typeof onCancel !== 'undefined') {
		returnValue.finally(() =>
			signal.removeEventListener('abort', onCancel)
		);
	}
	return returnValue;
}

function setIntervalPromise(after = 1, value, options = {}) {
	/* eslint-disable no-undefined, no-unreachable-loop, no-loop-func */
	try {
		validateObject(options, 'options');
	} catch (error) {
		return asyncIterator({
			next: function () {
				return Promise.reject(error);
			}
		});
	}
	const { signal, ref: reference = true } = options;
	try {
		validateAbortSignal(signal, 'options.signal');
	} catch (error) {
		return asyncIterator({
			next: function () {
				return Promise.reject(error);
			}
		});
	}
	try {
		validateBoolean(reference, 'options.ref');
	} catch (error) {
		return asyncIterator({
			next: function () {
				return Promise.reject(error);
			}
		});
	}
	if (signal?.aborted) {
		return asyncIterator({
			next: function () {
				return Promise.reject(new AbortError());
			}
		});
	}

	let onCancel, interval;

	try {
		let notYielded = 0;
		let callback;
		interval = setInterval(() => {
			notYielded++;
			if (callback) {
				callback();
				callback = undefined;
			}
		}, after);
		if (!reference) {
			interval?.unref?.();
		}
		if (signal) {
			onCancel = () => {
				clearInterval(interval);
				if (callback) {
					callback();
					callback = undefined;
				}
			};
			signal.addEventListener('abort', onCancel);
		}

		return asyncIterator({
			next: function () {
				return new Promise((resolve, reject) => {
					if (!signal?.aborted) {
						if (notYielded === 0) {
							callback = resolve;
						} else {
							resolve();
						}
					} else if (notYielded === 0) {
						reject(new AbortError());
					} else {
						resolve();
					}
				}).then(() => {
					if (notYielded > 0) {
						notYielded = notYielded - 1;
						return { done: false, value: value };
					}
					return { done: true };
				});
			},
			return: function () {
				clearInterval(interval);
				signal?.removeEventListener('abort', onCancel);
				return Promise.resolve({});
			}
		});
	} catch (error) {
		return asyncIterator({
			next: function () {
				clearInterval(interval);
				signal?.removeEventListener('abort', onCancel);
			}
		});
	}
}

export {
	setTimeoutPromise as setTimeout,
	setImmediatePromise as setImmediate,
	setIntervalPromise as setInterval
};


karma.conf.js
'use strict';

const path = require('path');
const fs = require('fs');
const { default: resolve } = require('@rollup/plugin-node-resolve');
const commonjs = require('@rollup/plugin-commonjs');
const nodeBuiltins = require('rollup-plugin-node-builtins');
const globals = require('rollup-plugin-node-globals');
const { default: babel } = require('@rollup/plugin-babel');
const istanbul = require('rollup-plugin-istanbul');
const rollupConfig = require('./rollup.config');

let config;

const isCI =
	typeof process.env.CI !== 'undefined' && process.env.CI !== 'false';
const isPR =
	typeof process.env.TRAVIS_PULL_REQUEST !== 'undefined' &&
	process.env.TRAVIS_PULL_REQUEST !== 'false';
const local = !isCI || (isCI && isPR);

const port = 0;

if (local) {
	config = {
		browsers: ['Chrome']
	};
} else {
	config = {
		hostname: 'bs-local.com',
		browserStack: {
			username: process.env.BROWSER_STACK_USERNAME,
			accessKey: process.env.BROWSER_STACK_ACCESS_KEY,
			startTunnel: true,
			project: 'isomorphic-timers-promises',
			name: 'Automated (Karma)',
			build: 'Automated (Karma)'
		},
		customLaunchers: {
			'BS-Chrome': {
				'base': 'BrowserStack',
				'project': 'isomorphic-timers-promises',
				'build': 'Automated (Karma)',
				'browser': 'Chrome',
				'browser_version': '72',
				'name': 'Chrome',
				'os': 'Windows',
				'os_version': '7'
			},
			'BS-Firefox': {
				'base': 'BrowserStack',
				'project': 'isomorphic-timers-promises',
				'build': 'Automated (Karma)',
				'browser': 'Firefox',
				'browser_version': '65',
				'name': 'Firefox',
				'os': 'Windows',
				'os_version': '7'
			},
			'BS-IE': {
				'base': 'BrowserStack',
				'project': 'isomorphic-timers-promises',
				'build': 'Automated (Karma)',
				'browser': 'IE',
				'browser_version': '11',
				'name': 'IE',
				'os': 'Windows',
				'os_version': '7'
			}
		},
		browsers: ['BS-Chrome', 'BS-Firefox', 'BS-IE']
	};
}

module.exports = function (baseConfig) {
	baseConfig.set({
		basePath: '',
		frameworks: ['mocha', 'fixture'],
		files: ['test/**/*.html', { pattern: 'test/**/*.js', watched: false }],
		exclude: [],
		preprocessors: {
			'test/**/*.html': ['html2js'],
			'test/**/*.js': ['rollup', 'sourcemap']
		},
		reporters: ['mocha', 'coverage-istanbul'],
		port: port,
		colors: true,
		logLevel: baseConfig.LOG_INFO,
		autoWatch: false,
		client: {
			captureConsole: true
		},
		browserConsoleLogOptions: {
			level: 'log',
			format: '%b %T: %m',
			terminal: true
		},
		rollupPreprocessor: {
			plugins: [
				nodeBuiltins(),
				babel({
					exclude: 'node_modules/**',
					babelHelpers: 'runtime'
				}),
				resolve({
					preferBuiltins: true
				}),
				commonjs(),
				babel({
					include:
						'node_modules/{has-flag,supports-color,set-immediate-shim}/**',
					babelHelpers: 'runtime',
					babelrc: false,
					configFile: path.resolve(__dirname, '.babelrc')
				}),
				globals(),
				...rollupConfig.plugins.filter(
					({ name }) => !['babel'].includes(name)
				),
				istanbul({
					exclude: ['test/**/*.js', 'node_modules/**/*']
				})
			],
			output: {
				format: 'iife',
				name: 'isomorphicTimersPromises',
				sourcemap: baseConfig.autoWatch ? false : 'inline', // Source map support has weird behavior in watch mode
				intro: 'window.TYPED_ARRAY_SUPPORT = false;' // IE9
			}
		},
		coverageIstanbulReporter: {
			dir: path.join(__dirname, 'coverage/%browser%'),
			fixWebpackSourcePaths: true,
			reports: ['html', 'text'],
			thresholds: {
				global: JSON.parse(
					fs.readFileSync(path.join(__dirname, '.nycrc'), 'utf8')
				)
			}
		},
		singleRun: true,
		concurrency: Infinity,
		...config
	});
};
